File: examples/opencog_agent_example.py
"""
Example: Using OpenCog-inspired Cognitive Agent with LlamaCloud Services

This example demonstrates how to use the CognitiveAgent to create an
autonomous agent with goals, beliefs, and reasoning capabilities.
"""

from llama_cloud_services import CognitiveAgent


def basic_cognitive_agent_example():
    """Basic example of creating and using a cognitive agent."""
    print("=" * 60)
    print("Basic Cognitive Agent Example")
    print("=" * 60)

    # Create a cognitive agent
    agent = CognitiveAgent(name="assistant_agent")

    # Define goals for the agent
    print("\n1. Setting agent goals...")
    agent.add_goal("help user with Python", priority=0.9)
    agent.add_goal("learn new information", priority=0.7)
    agent.add_goal("improve responses", priority=0.6)

    # Add available actions
    print("2. Registering available actions...")
    agent.add_action("search documentation", success_prob=0.8)
    agent.add_action("provide code example", success_prob=0.7)
    agent.add_action("ask clarifying question", success_prob=0.9)

    # Process observations
    print("3. Processing observations...")
    observations = [
        {"concept": "user needs help with Python functions", "strength": 0.9, "confidence": 0.9},
        {"concept": "user is a beginner", "strength": 0.7, "confidence": 0.8},
        {"concept": "documentation is available", "strength": 0.8, "confidence": 0.9},
    ]
    perception_result = agent.perceive(observations)
    print(f"   Perceived {perception_result['count']} concepts")

    # Run reasoning process
    print("4. Running reasoning process...")
    reasoning_result = agent.reason()
    print(f"   Considered {reasoning_result['beliefs_considered']} beliefs")
    print(f"   Evaluated {reasoning_result['goals_considered']} goals")

    # Plan actions
    print("5. Planning actions...")
    action_result = agent.plan_actions()
    print(f"   Selected {len(action_result['selected_actions'])} actions")
    for action in action_result["selected_actions"]:
        print(f"   - {action['action']} (score: {action['score']:.2f})")

    # Get knowledge summary
    print("\n6. Agent knowledge summary:")
    summary = agent.get_knowledge_summary()
    print(f"   Total atoms: {summary['total_atoms']}")
    print(f"   Goals: {summary['goals']}")
    print(f"   Actions: {summary['actions']}")
    print(f"   Implications: {summary['implications']}")


def cognitive_cycle_example():
    """Example of running a complete cognitive cycle."""
    print("\n" + "=" * 60)
    print("Cognitive Cycle Example")
    print("=" * 60)

    agent = CognitiveAgent(name="learning_agent")

    # Setup
    agent.add_goal("understand user intent", priority=0.9)
    agent.add_goal("provide helpful response", priority=0.8)
    agent.add_action("analyze input", success_prob=0.85)
    agent.add_action("generate response", success_prob=0.75)

    # Run cognitive cycle with new observations
    observations = [
        {"concept": "user asking about lists", "strength": 0.9},
        {"concept": "user confused about syntax", "strength": 0.7},
    ]

    print("\nRunning cognitive cycle...")
    result = agent.cognitive_cycle(observations)

    print(f"\nAgent: {result['agent']}")
    print("\nCycle results:")
    for process_name, process_result in result["cycle_results"].items():
        print(f"  {process_name}:")
        if isinstance(process_result, dict):
            for key, value in process_result.items():
                if key != "process":
                    print(f"    - {key}: {value}")


def knowledge_graph_example():
    """Example of working with the AtomSpace knowledge graph."""
    print("\n" + "=" * 60)
    print("Knowledge Graph Example")
    print("=" * 60)

    agent = CognitiveAgent(name="knowledge_agent")

    # Build knowledge graph
    print("\n1. Building knowledge graph...")
    
    # Add concepts using AtomType
    from llama_cloud_services.opencog.atomspace import AtomType
    
    concept1 = agent.atomspace.add_atom(
        atom_type=AtomType.CONCEPT,
        name="Python programming"
    )
    
    # Add beliefs
    belief1 = agent.add_belief("Python is beginner-friendly", strength=0.9)
    belief2 = agent.add_belief("Python has great libraries", strength=0.95)
    
    # Add goals
    goal1 = agent.add_goal("teach Python effectively", priority=0.9)
    
    print(f"   Added beliefs, goals, and concepts")

    # Query knowledge graph
    print("\n2. Querying knowledge graph...")
    
    beliefs = agent.atomspace.find_atoms(atom_type=AtomType.BELIEF)
    print(f"   Found {len(beliefs)} beliefs:")
    for belief in beliefs:
        print(f"   - {belief.name} (strength: {belief.truth_value.strength:.2f})")

    goals = agent.atomspace.find_atoms(atom_type=AtomType.GOAL)
    print(f"\n   Found {len(goals)} goals:")
    for goal in goals:
        print(f"   - {goal.name} (priority: {goal.truth_value.strength:.2f})")


def autonomous_agent_scenario():
    """Example of an autonomous agent scenario."""
    print("\n" + "=" * 60)
    print("Autonomous Agent Scenario")
    print("=" * 60)

    # Create an agent with specific purpose
    agent = CognitiveAgent(name="tutorial_assistant")

    # Define agent's goals
    agent.add_goal("help user learn Python", priority=0.95)
    agent.add_goal("provide clear examples", priority=0.85)
    agent.add_goal("maintain user engagement", priority=0.75)

    # Define agent's capabilities
    agent.add_action("provide code example", success_prob=0.9)
    agent.add_action("explain concept", success_prob=0.85)
    agent.add_action("suggest practice exercise", success_prob=0.8)
    agent.add_action("recommend resources", success_prob=0.75)

    # Initial beliefs
    agent.add_belief("users learn best by doing", strength=0.9)
    agent.add_belief("examples clarify concepts", strength=0.85)

    print("\nAgent initialized with goals, actions, and beliefs")

    # Simulate interaction over multiple cycles
    scenarios = [
        {
            "name": "User asks about functions",
            "observations": [
                {"concept": "user interested in functions", "strength": 0.9},
                {"concept": "user needs examples", "strength": 0.8},
            ],
        },
        {
            "name": "User practices code",
            "observations": [
                {"concept": "user writing code", "strength": 0.85},
                {"concept": "user making progress", "strength": 0.7},
            ],
        },
        {
            "name": "User needs clarification",
            "observations": [
                {"concept": "user confused", "strength": 0.8},
                {"concept": "concept needs explanation", "strength": 0.9},
            ],
        },
    ]

    for i, scenario in enumerate(scenarios, 1):
        print(f"\n--- Cycle {i}: {scenario['name']} ---")
        result = agent.cognitive_cycle(scenario["observations"])

        # Show selected actions
        actions = result["cycle_results"]["action_planning"]["selected_actions"]
        if actions:
            print(f"Agent selects: {actions[0]['action']}")
            print(f"  For goal: {actions[0]['goal']}")
            print(f"  Confidence: {actions[0]['score']:.2f}")
        else:
            print("Agent: No action needed at this time")


def main():
    """Run all examples."""
    print("\n" + "=" * 60)
    print("OpenCog Cognitive Agent Examples")
    print("LlamaCloud Services Integration")
    print("=" * 60)

    try:
        basic_cognitive_agent_example()
        cognitive_cycle_example()
        knowledge_graph_example()
        autonomous_agent_scenario()

        print("\n" + "=" * 60)
        print("All examples completed successfully!")
        print("=" * 60)

    except Exception as e:
        print(f"\nError running examples: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()

File: examples-ts/opencog-agent-example.ts
/**
 * Example: Using OpenCog-inspired Cognitive Agent with LlamaCloud Services
 * 
 * This example demonstrates how to use the CognitiveAgent to create an
 * autonomous agent with goals, beliefs, and reasoning capabilities.
 */

import {
  CognitiveAgent,
  AtomSpace,
  AtomType,
} from "llama-cloud-services";

/**
 * Basic example of creating and using a cognitive agent.
 */
function basicCognitiveAgentExample() {
  console.log("=".repeat(60));
  console.log("Basic Cognitive Agent Example");
  console.log("=".repeat(60));

  // Create a cognitive agent
  const agent = new CognitiveAgent("assistant_agent");

  // Define goals for the agent
  console.log("\n1. Setting agent goals...");
  agent.addGoal("help user with Python", 0.9);
  agent.addGoal("learn new information", 0.7);
  agent.addGoal("improve responses", 0.6);

  // Add available actions
  console.log("2. Registering available actions...");
  agent.addAction("search documentation", 0.8);
  agent.addAction("provide code example", 0.7);
  agent.addAction("ask clarifying question", 0.9);

  // Process observations
  console.log("3. Processing observations...");
  const observations = [
    { concept: "user needs help with Python functions", strength: 0.9, confidence: 0.9 },
    { concept: "user is a beginner", strength: 0.7, confidence: 0.8 },
    { concept: "documentation is available", strength: 0.8, confidence: 0.9 },
  ];
  const perceptionResult = agent.perceive(observations);
  console.log(`   Perceived ${perceptionResult.count} concepts`);

  // Run reasoning process
  console.log("4. Running reasoning process...");
  const reasoningResult = agent.reason();
  console.log(`   Considered ${reasoningResult.beliefsConsidered} beliefs`);
  console.log(`   Evaluated ${reasoningResult.goalsConsidered} goals`);

  // Plan actions
  console.log("5. Planning actions...");
  const actionResult = agent.planActions();
  console.log(`   Selected ${actionResult.selectedActions.length} actions`);
  for (const action of actionResult.selectedActions) {
    console.log(`   - ${action.action} (score: ${action.score.toFixed(2)})`);
  }

  // Get knowledge summary
  console.log("\n6. Agent knowledge summary:");
  const summary = agent.getKnowledgeSummary();
  console.log(`   Total atoms: ${summary.totalAtoms}`);
  console.log(`   Goals: ${summary.goals}`);
  console.log(`   Actions: ${summary.actions}`);
  console.log(`   Implications: ${summary.implications}`);
}

/**
 * Example of running a complete cognitive cycle.
 */
function cognitiveCycleExample() {
  console.log("\n" + "=".repeat(60));
  console.log("Cognitive Cycle Example");
  console.log("=".repeat(60));

  const agent = new CognitiveAgent("learning_agent");

  // Setup
  agent.addGoal("understand user intent", 0.9);
  agent.addGoal("provide helpful response", 0.8);
  agent.addAction("analyze input", 0.85);
  agent.addAction("generate response", 0.75);

  // Run cognitive cycle with new observations
  const observations = [
    { concept: "user asking about lists", strength: 0.9 },
    { concept: "user confused about syntax", strength: 0.7 },
  ];

  console.log("\nRunning cognitive cycle...");
  const result = agent.cognitiveCycle(observations);

  console.log(`\nAgent: ${result.agent}`);
  console.log("\nCycle results:");
  for (const [processName, processResult] of Object.entries(result.cycleResults)) {
    console.log(`  ${processName}:`);
    if (typeof processResult === "object" && processResult !== null) {
      for (const [key, value] of Object.entries(processResult)) {
        if (key !== "process") {
          console.log(`    - ${key}: ${JSON.stringify(value)}`);
        }
      }
    }
  }
}

/**
 * Example of working with the AtomSpace knowledge graph.
 */
function knowledgeGraphExample() {
  console.log("\n" + "=".repeat(60));
  console.log("Knowledge Graph Example");
  console.log("=".repeat(60));

  const agent = new CognitiveAgent("knowledge_agent");

  // Build knowledge graph
  console.log("\n1. Building knowledge graph...");

  // Add concepts using AtomType
  const concept1 = agent.atomspace.addAtom(
    AtomType.CONCEPT,
    "Python programming"
  );

  // Add beliefs
  const belief1 = agent.addBelief("Python is beginner-friendly", 0.9);
  const belief2 = agent.addBelief("Python has great libraries", 0.95);

  // Add goals
  const goal1 = agent.addGoal("teach Python effectively", 0.9);

  console.log("   Added beliefs, goals, and concepts");

  // Query knowledge graph
  console.log("\n2. Querying knowledge graph...");

  const beliefs = agent.atomspace.findAtoms({ atomType: AtomType.BELIEF });
  console.log(`   Found ${beliefs.length} beliefs:`);
  for (const belief of beliefs) {
    console.log(`   - ${belief.name} (strength: ${belief.truthValue.strength.toFixed(2)})`);
  }

  const goals = agent.atomspace.findAtoms({ atomType: AtomType.GOAL });
  console.log(`\n   Found ${goals.length} goals:`);
  for (const goal of goals) {
    console.log(`   - ${goal.name} (priority: ${goal.truthValue.strength.toFixed(2)})`);
  }
}

/**
 * Example of an autonomous agent scenario.
 */
function autonomousAgentScenario() {
  console.log("\n" + "=".repeat(60));
  console.log("Autonomous Agent Scenario");
  console.log("=".repeat(60));

  // Create an agent with specific purpose
  const agent = new CognitiveAgent("tutorial_assistant");

  // Define agent's goals
  agent.addGoal("help user learn Python", 0.95);
  agent.addGoal("provide clear examples", 0.85);
  agent.addGoal("maintain user engagement", 0.75);

  // Define agent's capabilities
  agent.addAction("provide code example", 0.9);
  agent.addAction("explain concept", 0.85);
  agent.addAction("suggest practice exercise", 0.8);
  agent.addAction("recommend resources", 0.75);

  // Initial beliefs
  agent.addBelief("users learn best by doing", 0.9);
  agent.addBelief("examples clarify concepts", 0.85);

  console.log("\nAgent initialized with goals, actions, and beliefs");

  // Simulate interaction over multiple cycles
  const scenarios = [
    {
      name: "User asks about functions",
      observations: [
        { concept: "user interested in functions", strength: 0.9 },
        { concept: "user needs examples", strength: 0.8 },
      ],
    },
    {
      name: "User practices code",
      observations: [
        { concept: "user writing code", strength: 0.85 },
        { concept: "user making progress", strength: 0.7 },
      ],
    },
    {
      name: "User needs clarification",
      observations: [
        { concept: "user confused", strength: 0.8 },
        { concept: "concept needs explanation", strength: 0.9 },
      ],
    },
  ];

  scenarios.forEach((scenario, i) => {
    console.log(`\n--- Cycle ${i + 1}: ${scenario.name} ---`);
    const result = agent.cognitiveCycle(scenario.observations);

    // Show selected actions
    const actions = result.cycleResults.actionPlanning?.selectedActions || [];
    if (actions.length > 0) {
      console.log(`Agent selects: ${actions[0].action}`);
      console.log(`  For goal: ${actions[0].goal}`);
      console.log(`  Confidence: ${actions[0].score.toFixed(2)}`);
    } else {
      console.log("Agent: No action needed at this time");
    }
  });
}

/**
 * Run all examples.
 */
function main() {
  console.log("\n" + "=".repeat(60));
  console.log("OpenCog Cognitive Agent Examples");
  console.log("LlamaCloud Services Integration");
  console.log("=".repeat(60));

  try {
    basicCognitiveAgentExample();
    cognitiveCycleExample();
    knowledgeGraphExample();
    autonomousAgentScenario();

    console.log("\n" + "=".repeat(60));
    console.log("All examples completed successfully!");
    console.log("=".repeat(60));
  } catch (error) {
    console.error("\nError running examples:", error);
    if (error instanceof Error) {
      console.error(error.stack);
    }
  }
}

// Run if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export {
  basicCognitiveAgentExample,
  cognitiveCycleExample,
  knowledgeGraphExample,
  autonomousAgentScenario,
  main,
};
